c7n/resources/ec2.py

```diff
@@ -1476,31 +1482,66 @@ def process(self, instances):
-                [i['InstanceId'] for i in ephemeral])
+            self._run_instances_op(client, 'terminate', ephemeral)
         if persistent:
             if self.data.get('hibernate', False):
                 enabled, persistent = self.split_on_hibernate(persistent)
                 if enabled:
-                    self._run_instances_op(
-                        client.stop_instances,
-                        [i['InstanceId'] for i in enabled],
-                        Hibernate=True)
-            self._run_instances_op(
-                client.stop_instances,
-                [i['InstanceId'] for i in persistent])
+                    self._run_instances_op(client, 'stop', enabled, Hibernate=True)
+            self._run_instances_op(client, 'stop', persistent)
         return instances
 
-    def _run_instances_op(self, op, instance_ids, **kwargs):
-        while instance_ids:
+    def disable_protection(self, client, op, instances):
+        def modify_instance(i, attribute):
             try:
-                return self.manager.retry(op, InstanceIds=instance_ids, **kwargs)
+                self.manager.retry(
+                    client.modify_instance_attribute,
+                    InstanceId=i['InstanceId'],
+                    Attribute=attribute,
+                    Value='false',
+                )
             except ClientError as e:
                 if e.response['Error']['Code'] == 'IncorrectInstanceState':
-                    instance_ids.remove(extract_instance_id(e))
+                    return
                 raise
 
+        def process_instance(i, op):
+            modify_instance(i, 'disableApiStop')
+            if op == 'terminate':
+                modify_instance(i, 'disableApiTermination')
+
+        with self.executor_factory(max_workers=2) as w:
+            list(w.map(process_instance, instances, op))
+
+    def _run_instances_op(self, client, op, instances, **kwargs):
+        while instances:
+
+            client_op = client.stop_instances
+            if op == 'terminate':
+                client_op = client.terminate_instances
+
+            instance_ids = [i['InstanceId'] for i in instances]
+
+            try:
+                return self.manager.retry(client_op, InstanceIds=instance_ids, **kwargs)
+            except ClientError as e:
+                if e.response['Error']['Code'] == 'IncorrectInstanceState':
+                    instance_ids.remove(extract_instance_id(e))
+                if e.response['Error']['Code'] != 'OperationNotPermitted':
+                    raise
+                if not self.data.get('force'):
+                    raise
+
+                self.log.info("Disabling stop and termination protection on instances")
+                self.disable_protection(
+                    client,
+                    op,
+                    [i for i in instances if i.get('InstanceLifecycle') != 'spot'],
+                )
+                return self.manager.retry(
```

Comment:
should this be here, we'll go back into the next iteration of the loop to try client_op

Reply:
Yes, I edited it. Is it okay ?